File: src/content/docs/topics/cs/leetcode-150/1d-dynamic-programming/152-maximum-product-subarray.py
def max_product(nums):
    pass  # TODO: implement

def _run_tests():
    assert max_product([2, 3, -2, 4]) == 6
    assert max_product([-2, 0, -1]) == 0
    assert max_product([-2, 3, -4]) == 24
    assert max_product([0]) == 0
    assert max_product([-3]) == -3
    assert max_product([-2, -3]) == 6
    # --- large-input timing ---
    import time as _t
    _t0 = _t.perf_counter()
    max_product(list(range(500, 0, -1)))
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf max_product(range 500 desc): {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/1d-dynamic-programming/152-maximum-product-subarray.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function maxProduct(nums: number[]): number {
    // TODO: implement
    return 0;
}

function _runTests(): void {
    assert(maxProduct([2, 3, -2, 4]) === 6);
    assert(maxProduct([-2, 0, -1]) === 0);
    assert(maxProduct([-2, 3, -4]) === 24);
    assert(maxProduct([0]) === 0);
    assert(maxProduct([-3]) === -3);
    assert(maxProduct([-2, -3]) === 6);
    // perf
    const t0 = performance.now();
    maxProduct(Array.from({ length: 500 }, (_, i) => (i % 3 === 0 ? -1 : i + 1)));
    console.log(`perf maxProduct(500 mixed): ${(performance.now() - t0).toFixed(1)}ms`);
    console.log('all tests pass');
}

_runTests();
